src/Cli/dotnet/commands/VirtualProjectBuildingCommand.cs

```diff
@@ -0,0 +1,192 @@
+            if (buildResult.OverallResult != BuildResultCode.Success)
+            {
+                return 1;
+            }
+
+            BuildManager.DefaultBuildManager.EndBuild();
+            return 0;
+        }
+        catch (Exception e)
+        {
+            Console.Error.WriteLine(e.Message);
+            return 1;
+        }
+        finally
+        {
+            foreach (var (key, value) in savedEnvironmentVariables)
+            {
+                Environment.SetEnvironmentVariable(key, value);
+            }
+
+            binaryLogger?.Shutdown();
+            consoleLogger.Shutdown();
+        }
+
+        static ILogger? GetBinaryLogger(string[] args)
+        {
+            for (int i = args.Length - 1; i >= 0; i--)
+            {
+                var arg = args[i];
+                if (RunCommand.IsBinLogArgument(arg))
+                {
+                    return new BinaryLogger
+                    {
+                        Parameters = arg.IndexOf(':') is >= 0 and var index
+                            ? arg[(index + 1)..]
+                            : "msbuild.binlog",
+                    };
+                }
+            }
+
+            return null;
+        }
+    }
+
+    public ProjectInstance CreateProjectInstance(ProjectCollection projectCollection)
+    {
+        return CreateProjectInstance(projectCollection, addGlobalProperties: null);
+    }
+
+    private ProjectInstance CreateProjectInstance(
+        ProjectCollection projectCollection,
+        Action<IDictionary<string, string>>? addGlobalProperties = null)
+    {
+        var projectRoot = CreateProjectRootElement(projectCollection);
+
+        var globalProperties = projectCollection.GlobalProperties;
+        if (addGlobalProperties is not null)
+        {
+            globalProperties = new Dictionary<string, string>(projectCollection.GlobalProperties, StringComparer.OrdinalIgnoreCase);
+            addGlobalProperties(globalProperties);
+        }
+
+        return ProjectInstance.FromProjectRootElement(projectRoot, new ProjectOptions
+        {
+            GlobalProperties = globalProperties,
+        });
+    }
+
+    private ProjectRootElement CreateProjectRootElement(ProjectCollection projectCollection)
+    {
+        var projectFileFullPath = Path.ChangeExtension(EntryPointFileFullPath, ".csproj");
+        var projectFileText = """
+            <Project>
+                <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
+
+                <PropertyGroup>
+                    <OutputType>Exe</OutputType>
+                    <TargetFramework>net9.0</TargetFramework>
+                    <ImplicitUsings>enable</ImplicitUsings>
+                    <Nullable>enable</Nullable>
+                </PropertyGroup>
+
+                <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />
+
+                <!-- Override targets which don't work with project files that are not present on disk. -->
```

Comment:
Ok now this is some deep magic. It looks like all of this is Restore-time stuff, do we need to add some kinds of flags/detections to enable in-memory use cases for real instead of requiring these workarounds?

cc @rainersigwald / @nkolev92

Reply:
> It feels strange that an xml string is needed to make an in-memory project, when we are using the msbuild APIs directly for so many other things.

Isn't that same with strings containing C# that you need when using Roslyn APIs? You could create SyntaxTrees manually but that wouldn't be recommended. Similarly, I could create the MSBuild project without the string but seems that would be less readable.

Reply:
I think I am relatively comfortable with it, especially if it remains as "just a string" and we don't start using a complex interpolated string with XML escaping handling, etc. to make the project.

Reply:
> Ok now this is some deep magic. It looks like all of this is Restore-time stuff, do we need to add some kinds of flags/detections to enable in-memory use cases for real instead of requiring these workarounds?

Yes, ideally the targets would allow in-memory builds. Should I file an issue for it at https://github.com/NuGet/Home since the targets live there I think?

Reply:
Yes please - the NuGet team is currently soliciting cross-team requests for the next semester, so this would be a good candidate to ask for.

Reply:
Yes please - the NuGet team is currently soliciting cross-team requests for the next semester, so this would be a good candidate to ask for.

Reply:
Filed https://github.com/NuGet/Home/issues/14148.

Reply:
We'll probably need to have a meeting to discuss this.  Restore itself is one hurdle but also NuGet writes out files `project.assets.json`, `project.nuget.g.props`, and `project.nuget.g.targets`.  The `.props` and `.targets` are imported by the SDK using a file pattern:

https://github.com/dotnet/msbuild/blob/41ecead49820b70e3ae6526a2ba0bec6cdcbb934/src/Tasks/Microsoft.Common.props#L74

And the `project.assets.json` is loaded by the SDK during the build to add MSBuild items on the fly that represent things to pass around:

https://github.com/dotnet/sdk/blob/main/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets#L266-L313

Both of these are very dependent on files on disk.  I have potential workarounds but we should chat.

Reply:
These are in `obj` folder right? That still exists (I imagine it would be much more difficult to get rid of that), so I don't think this should be a problem.

Reply:
not quite - the big problem is that a key file in the `obj` folder is not 'scoped' to a single project - it's called `project.assets.json`. This name is customizable though, so your in-memory project here should set that property so that we're future-proof for the "multiple apps in one directory" scenario. That property name is `ProjectAssetsFile`.

You probably want to make sure that they synthetic project name is unique-per-app too to account for MSBuild logic that does things like `$(MSBuildProjectName).foo.bar`.

Reply:
Can you help me understand why is the scoping of the file name a problem?

@jjonescz I think it's a good to consider some sort of a design review for https://github.com/dotnet/sdk/pull/47020. 

It involves NuGet gestures, but the NuGet team hasn't really had a chance to look into that yet. 
I've tagged the appropriate people, but it's a bit of short notice for everyone to fully wrap their heads around what's going on.

Reply:
@nkolev92 when I was looking at the `ProjectAssetsFile` logic in the targets, right next to it I saw other logic that was using the `MSBuildProjectName` to compute an intermediate file path - to me that implies that we need to be careful to ensure each 'app' has a distinct project name so that intermediates could reasonably be created in the same intermediate directory without clobbering

Reply:
Scoping could be a problem if we allow multiple virtual projects in one folder, but the proposal currently says they would get separate artifacts folders, so that should avoid any problems.

> short notice

I think it is expected that the implementation details of this feature might change over time, so there is no rush.